File: CRM_tecnologia_backend/app/schemas/documento.py
from typing import Optional, List, Any
from datetime import datetime
# pyrefly: ignore [missing-import]
from pydantic import BaseModel, ConfigDict


class DocumentoBase(BaseModel):
    nombre: str
    tipo: str  # 'pdf', 'word', 'docx', 'doc'
    tamanio: str  # '2.4 MB'
    tamanio_bytes: Optional[int] = 0
    categoria: Optional[str] = "General"
    descripcion: Optional[str] = None
    archivo_url: Optional[str] = None
    archivo_base64: Optional[str] = None
    subido_por: str
    usuario_id: Optional[str] = None
    usuario_rol: Optional[str] = "analista"
    tags_json: Optional[List[str]] = None


class DocumentoCreate(DocumentoBase):
    pass


class DocumentoUpdate(BaseModel):
    nombre: Optional[str] = None
    categoria: Optional[str] = None
    descripcion: Optional[str] = None
    tags_json: Optional[List[str]] = None


class DocumentoResponse(BaseModel):
    id: str
    nombre: str
    tipo: str
    tamanio: str
    tamanio_bytes: int
    categoria: str
    descripcion: Optional[str] = None
    archivo_url: Optional[str] = None
    archivo_base64: Optional[str] = None
    subido_por: str
    usuario_id: Optional[str] = None
    usuario_rol: str
    tags_json: Optional[List[Any]] = None
    created_at: datetime
    updated_at: datetime

    model_config = ConfigDict(from_attributes=True)

File: CRM_tecnologia_frontend/src/types/documento.ts
export type DocumentoTipo = 'pdf' | 'word' | 'docx' | 'doc' | 'rtf' | 'odt' | 'otro';

export type DocumentoCategoria =
  | 'Contratos'
  | 'Reportes Ejecutivos'
  | 'Propuestas Comerciales'
  | 'Especificaciones Técnicas'
  | 'Financiero'
  | 'Auditoría'
  | 'General';

export interface Documento {
  id: string;
  nombre: string;
  tipo: DocumentoTipo | string;
  tamanio: string;
  tamanio_bytes: number;
  categoria: DocumentoCategoria | string;
  descripcion?: string | null;
  archivo_url?: string | null;
  archivo_base64?: string | null;
  subido_por: string;
  usuario_id?: string | null;
  usuario_rol: 'analista' | 'administrador' | string;
  tags_json?: string[] | null;
  /** Roles que pueden ver este documento. ["todos"] o subconjunto de los roles del sistema. */
  destinatarios_roles?: string[] | null;
  created_at: string;
  updated_at: string;
}

export interface DocumentoCreatePayload {
  nombre: string;
  tipo: DocumentoTipo | string;
  tamanio: string;
  tamanio_bytes?: number;
  categoria?: DocumentoCategoria | string;
  descripcion?: string;
  archivo_url?: string;
  archivo_base64?: string;
  subido_por: string;
  usuario_id?: string;
  usuario_rol?: 'analista' | 'administrador' | string;
  tags_json?: string[];
  destinatarios_roles?: string[];
}

export interface DocumentoUpdatePayload {
  nombre?: string;
  categoria?: DocumentoCategoria | string;
  descripcion?: string;
  tags_json?: string[];
}

export interface DocumentosContextType {
  documentos: Documento[];
  isLoading: boolean;
  isUploading: boolean;
  error: string | null;
  fetchDocumentos: () => Promise<void>;
  uploadDocumentoFile: (
    file: File,
    meta: {
      categoria?: DocumentoCategoria | string;
      descripcion?: string;
      tags?: string[];
      destinatarios_roles?: string[];
    }
  ) => Promise<{ success: boolean; documento?: Documento; error?: string }>;
  createDocumentoDirect: (
    payload: DocumentoCreatePayload
  ) => Promise<{ success: boolean; documento?: Documento; error?: string }>;
  updateDocumento: (
    id: string,
    payload: DocumentoUpdatePayload
  ) => Promise<{ success: boolean; error?: string }>;
  deleteDocumento: (id: string) => Promise<{ success: boolean; error?: string }>;
  downloadDocumento: (doc: Documento) => void;
  generateSampleDocuments: () => void;
}
